### ISSUE TITLE
Layout/TrailingWhitespace + Lint/LiteralInInterpolation = infinite loop when autocorrecting

### ISSUE THREAD

**Issue by xxx** [rubocop-hq/rubocop#8975]
Heredocs with trailing whitespace on any line leads to an infinite loop during autocorrect.

Note this is similar to, but not the same thing as, https://github.com/rubocop-hq/rubocop/issues/8944 (which is an interaction with `Lint/EmptyInterpolation` rather than `Lint/LiteralInInterpolation`)

--------

## Expected behavior

I expect rubocop to not loop infinitely when running into a heredoc that has any trailing whitespace

## Actual behavior

There's an infinite loop during autocorrect if a heredoc line has trailing whitespace.

Please use `rubocop --debug` when pasting rubocop output as it contains additional information.

```
Inspecting 1 file
Scanning /home/mpd/rubocop-test/test.rb
W

Offenses:

test.rb:5:63: C: [Corrected] Layout/TrailingWhitespace: Trailing whitespace detected.
    there is a single trailing whitespace after this sentence.
                                                              ^
test.rb:5:65: W: [Corrected] Lint/LiteralInInterpolation: Literal interpolation detected.
    there is a single trailing whitespace after this sentence.#{' '}
                                                                ^^^

0 files inspected, 2 offenses detected, 2 offenses corrected
Infinite loop detected in /home/mpd/rubocop-test/test.rb and caused by Layout/TrailingWhitespace -> Lint/LiteralInInterpolation
/home/mpd/.rvm/gems/ruby-2.7.1/gems/rubocop-1.1.0/lib/rubocop/runner.rb:304:in `check_for_infinite_loop'
/home/mpd/.rvm/gems/ruby-2.7.1/gems/rubocop-1.1.0/lib/rubocop/runner.rb:287:in `block in iterate_until_no_changes'
/home/mpd/.rvm/gems/ruby-2.7.1/gems/rubocop-1.1.0/lib/rubocop/runner.rb:286:in `loop'
/home/mpd/.rvm/gems/ruby-2.7.1/gems/rubocop-1.1.0/lib/rubocop/runner.rb:286:in `iterate_until_no_changes'
/home/mpd/.rvm/gems/ruby-2.7.1/gems/rubocop-1.1.0/lib/rubocop/runner.rb:255:in `do_inspection_loop'
/home/mpd/.rvm/gems/ruby-2.7.1/gems/rubocop-1.1.0/lib/rubocop/runner.rb:132:in `block in file_offenses'
/home/mpd/.rvm/gems/ruby-2.7.1/gems/rubocop-1.1.0/lib/rubocop/runner.rb:157:in `file_offense_cache'
/home/mpd/.rvm/gems/ruby-2.7.1/gems/rubocop-1.1.0/lib/rubocop/runner.rb:131:in `file_offenses'
/home/mpd/.rvm/gems/ruby-2.7.1/gems/rubocop-1.1.0/lib/rubocop/runner.rb:122:in `process_file'
/home/mpd/.rvm/gems/ruby-2.7.1/gems/rubocop-1.1.0/lib/rubocop/runner.rb:101:in `block in each_inspected_file'
/home/mpd/.rvm/gems/ruby-2.7.1/gems/rubocop-1.1.0/lib/rubocop/runner.rb:100:in `each'
/home/mpd/.rvm/gems/ruby-2.7.1/gems/rubocop-1.1.0/lib/rubocop/runner.rb:100:in `reduce'
/home/mpd/.rvm/gems/ruby-2.7.1/gems/rubocop-1.1.0/lib/rubocop/runner.rb:100:in `each_inspected_file'
/home/mpd/.rvm/gems/ruby-2.7.1/gems/rubocop-1.1.0/lib/rubocop/runner.rb:86:in `inspect_files'
/home/mpd/.rvm/gems/ruby-2.7.1/gems/rubocop-1.1.0/lib/rubocop/runner.rb:47:in `run'
/home/mpd/.rvm/gems/ruby-2.7.1/gems/rubocop-1.1.0/lib/rubocop/cli/command/execute_runner.rb:25:in `execute_runner'
/home/mpd/.rvm/gems/ruby-2.7.1/gems/rubocop-1.1.0/lib/rubocop/cli/command/execute_runner.rb:17:in `run'
/home/mpd/.rvm/gems/ruby-2.7.1/gems/rubocop-1.1.0/lib/rubocop/cli/command.rb:11:in `run'
/home/mpd/.rvm/gems/ruby-2.7.1/gems/rubocop-1.1.0/lib/rubocop/cli/environment.rb:18:in `run'
/home/mpd/.rvm/gems/ruby-2.7.1/gems/rubocop-1.1.0/lib/rubocop/cli.rb:65:in `run_command'
/home/mpd/.rvm/gems/ruby-2.7.1/gems/rubocop-1.1.0/lib/rubocop/cli.rb:72:in `execute_runners'
/home/mpd/.rvm/gems/ruby-2.7.1/gems/rubocop-1.1.0/lib/rubocop/cli.rb:41:in `run'
/home/mpd/.rvm/gems/ruby-2.7.1/gems/rubocop-1.1.0/exe/rubocop:13:in `block in <top (required)>'
/home/mpd/.rvm/rubies/ruby-2.7.1/lib/ruby/2.7.0/benchmark.rb:308:in `realtime'
/home/mpd/.rvm/gems/ruby-2.7.1/gems/rubocop-1.1.0/exe/rubocop:12:in `<top (required)>'
/home/mpd/.rvm/gems/ruby-2.7.1/bin/rubocop:23:in `load'
/home/mpd/.rvm/gems/ruby-2.7.1/bin/rubocop:23:in `<main>'
/home/mpd/.rvm/gems/ruby-2.7.1/bin/ruby_executable_hooks:24:in `eval'
/home/mpd/.rvm/gems/ruby-2.7.1/bin/ruby_executable_hooks:24:in `<main>'
Finished in 0.14869983028620481 seconds
```
## Steps to reproduce the problem

This is extremely important! Providing us with a reliable way to reproduce
a problem will expedite its solution.

This does it. Make sure the trailing whitespace is not lost before running rubocop.

```ruby
class Test
  FOO = <<~STRING
    there is a single trailing whitespace after this sentence. 
  STRING
end

```
## RuboCop version

Include the output of `rubocop -V` or `bundle exec rubocop -V` if using Bundler. Here's an example:

```
1.1.0 (using Parser 2.7.2.0, rubocop-ast 1.1.0, running on ruby 2.7.1 x86_64-linux)
```

**Comment by marcandre** [comment]
Agreed, ideally `Lint/LiteralInInterpolation` would be clever enough to detect this kind of interpolation... (space literal, end of line, in HEREDOC)

**Comment by jdashton** [comment]
This fix only partially resolved the issue. When additional cops get involved, the infinite loop still occurs.

Here is the error reported by Rubocop:
```
Infinite loop detected in /Users/jdashton/learn/aoc2022-ruby/spec/aoc2022/puzzles/supply_stacks_spec.rb and caused by Lint/LiteralInInterpolation -> Layout/TrailingWhitespace -> Layout/SpaceInsideStringInterpolation, Style/StringLiteralsInInterpolation
```

The heredoc that causes this loop looks like this:
```ruby
    subject(:supply_stacks) { described_class.new StringIO.new(<<~DATA) }
          [D]#{ "    " }
      [N] [C]#{ "    " }
      [Z] [M] [P]
       1   2   3#{ " " }

      move 1 from 2 to 1
      move 3 from 1 to 3
      move 2 from 2 to 1
      move 1 from 1 to 2
    DATA
```

This is with Rubocop 1.39.0:
```bash
$ bin/rubocop -V
1.39.0 (using Parser 3.1.3.0, rubocop-ast 1.24.0, running on ruby 3.2.0) [x86_64-darwin22]
  - rubocop-rake 0.6.0
  - rubocop-rspec 2.15.0
```

The project `.rubocop.yml` file contains these settings for two of the cops, and no settings for the other two:
```yaml
Layout/SpaceInsideStringInterpolation:
  EnforcedStyle: space

Style/StringLiteralsInInterpolation:
  Enabled: true
  EnforcedStyle: double_quotes
```
